File: 0424-longest-repeating-character-replacement/0424-longest-repeating-character-replacement.cpp
class Solution {
public:
    int characterReplacement(string s, int k) {
        int n = s.size();

        int maxLen = 0;
        map<char, int> freq;
        multiset<int> freqOrder;
        for (int beg = 0, end = 0; end < n; end++) {
            char ch = s[end];
            freq[ch]++;
            freqOrder.insert(freq[ch]);
            if (freqOrder.find(freq[ch] - 1) != freqOrder.end()) {
                freqOrder.erase(freqOrder.find(freq[ch] - 1));
            }

            while (beg <= end && (end - beg + 1) - (*(--freqOrder.end())) > k) {
                char del = s[beg];
                freq[del]--;
                freqOrder.insert(freq[del]);
                freqOrder.erase(freqOrder.find(freq[del] + 1));
                

                if (freq[del] == 0) {
                    freq.erase(del);

                    if (freqOrder.find(0) != freqOrder.end()) {
                        freqOrder.erase(freqOrder.find(0));
                    }
                }

                beg++;
            }

            maxLen = max(maxLen, end - beg + 1);
        }

        return maxLen;
    }
};

File: 0424-longest-repeating-character-replacement/0424-longest-repeating-character-replacement.java
class Solution {
    public int characterReplacement(String s, int k) {
        char[] sarr = s.toCharArray();
        int n = sarr.length;

        int head = -1, tail = 0;
        int[] freq = new int[26];
        int maxFreq = 0;
        int ans = 0;
        while (tail < n) {
            while (head + 1 < n) {
                head += 1;
                char ch = sarr[head];

                freq[ch - 'A'] += 1;
                maxFreq = Math.max(maxFreq, freq[ch - 'A']);
                int replaceable = (head - tail + 1) - maxFreq;

                if (replaceable > k) {
                    head -= 1;
                    freq[ch - 'A'] -= 1;
                    for (int i = 0; i < 26; i++) {
                        maxFreq = Math.max(maxFreq, freq[i]);
                    }
                    break;
                }
            }

            ans = Math.max(ans, (head - tail + 1));

            if (head < tail) {
                tail += 1;
                head = tail - 1;
            } else {
                char ch = sarr[tail];
                freq[ch - 'A'] -= 1;
                for (int i = 0; i < 26; i++) {
                    maxFreq = Math.max(maxFreq, freq[i]);
                }
 
                tail += 1;
            }
        }
        
        return ans;   
    }
}